anndata/_core/ref_path.py

```diff
@@ -0,0 +1,387 @@
+from enum import Enum
+from itertools import product
+from typing import Union, Optional  # Special
+from typing import Iterable, Sequence, Generator  # ABCs
+from typing import Dict, Tuple  # Classes
+
+import pandas as pd
+import numpy as np
+import scipy.sparse as ssp
+
+from ..compat import Literal
+from . import anndata
+from .index import get_x_vector
+from .raw import Raw
+
+
+# TODO: allow sequences: ("obsm", "X_pca", [0, 1])
+from ..utils import _doc_params
+
+
+class AttrInfo(Enum):
+    layers = ((str, ("obs", "var"), str), dict(l=(), X=("X",)))
+    obs = ((str,), dict(o=()))
+    obsm = ((str, (int, str)), dict(om=()))
+    obsp = ((str, str, (0, 1)), dict(op=()))
+    var = ((str,), dict(v=()))
+    varm = ((str, (int, str)), dict(vm=()))
+    varp = ((str, str, (0, 1)), dict(vp=()))
+    raw = ((("X", "var", "varm"), ...), dict(rX=("X",), rv=("var",), rvm=("varm",)))
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}.{self.name}"
+
+    @property
+    def validation(self):
+        return self.value[0]
+
+    @property
+    def short_codes(self):
+        return self.value[1]
+
+    def validate(self, path: Sequence[Union[str, int]]):
+        if self is not AttrInfo.raw and not len(path) == len(self.validation):
+            raise ValueError(
+                f"Path length mismatch: required ({len(self.validation)}) ≠ "
+                f"got ({len(path)}) in path {path!r} for attr {self.name}."
+            )
+        elif self is AttrInfo.raw:
+            sub_path = RefPath.parse(path)
+            # layers or obs
+            sub_path._attr.validate(sub_path.path)
+            return
+
+        for i, (elem, check) in enumerate(zip(path, self.validation)):
+            err_prefix = f"Element path[{i}]={path[i]!r} in path {path!r}"
+            if isinstance(check, tuple):
+                if isinstance(check[0], type):
+                    if not any(isinstance(elem, c) for c in check):
+                        raise ValueError(
+                            f"{err_prefix} is not of one of the types {check!r}"
+                        )
+                elif not any(elem == c for c in check):
+                    raise ValueError(f"{err_prefix} is not of one of {check!r}")
+            elif isinstance(check, type):
+                if not isinstance(elem, check):
+                    raise ValueError(f"{err_prefix} is non of type {check}")
+            else:
+                assert False, f"Unhandled check {check!r} for {self}"
```

Comment:
Could you post some examples of what how these error could come up and what they'd look like? Mostly for my interest.

Reply:
It’s just when you try to construct an invalid path (too long, elements of the wrong type or value)

See `test_parse_failures`.